fix: When using Slack V2, file is changed to BytesIO

JaeGeunBang opened the pull request:
<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
fix(dashboard): load charts correctly
-->

### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

I add ALERT_REPORT_SLACK_V2: True in FEATURE_FLAGS.
```
FEATURE_FLAGS = {
      "ALERT_REPORT_SLACK_V2": True
    }
```


After switching to the Slack V2 API, I'm now sending .csv files via Superset Reports. The following issue has occurred. Slack should display the data as CSV, but it's being displayed as plain text.

Note that this issue didn't occur in Slack V1. Please see the screenshot below.

So, I discovered the issue in the Slack V2 code, fixed it, and confirmed that the CSV file is now displayed properly in Slack.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

Slack V1
<img width="695" height="680" alt="스크린샷 2025-11-13 오후 12 36 27" src="https://github.com/user-attachments/assets/7ad609ec-9772-4df2-a7c8-10aa37fd0f0d" />


Slack V2
<img width="631" height="827" alt="스크린샷 2025-11-13 오후 12 36 34" src="https://github.com/user-attachments/assets/61266231-4744-4d80-adb7-7920172e2767" />


### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

tw-addy-b committed:
fix: When using Slack V2, file is changed to BytesIO

gabotorresruiz commented:
> @gabotorresruiz you've been digging into the slack code a bit, have you been seeing this too?

Hmm I haven't tested this, but I can definitely take a look

JaeGeunBang committed:
test: add regression coverage for Slack V2 BytesIO file upload

Covers the CSV-as-plain-text bug where files_upload_v2 received raw
bytes instead of a file-like object. Also updates existing upload
assertions that compared the file kwarg against raw bytes, which broke
once the payload is wrapped in io.BytesIO.

JaeGeunBang committed:
fix: narrow _get_inline_files return type to satisfy mypy

files_upload_v2 wraps the returned file content in io.BytesIO, which
only accepts bytes-like buffers. NotificationContent's csv/xlsx/pdf/
screenshots fields are always bytes, so narrow the return type from
Sequence[Union[str, IOBase, bytes]] to Sequence[bytes] instead of
loosening the BytesIO call.

rusackas reviewed (approved):
The `io.BytesIO` wrap plus the regression test cover the plain-text bug well. LGTM.

rusackas committed:
Merge branch 'master' into fix-slack-v2-by-csv-file

rusackas reviewed (changes requested):
@JaeGeunBang welp, digging into this more after approving and I don't think it's safe as-is, so pulling my approval for now, sorry for the back-and-forth.

`_call_slack_api` retries transient `SlackApiError`s via `@backoff.on_exception`, and `files_upload_v2` reads the file stream on every call. `file_obj` is a single `io.BytesIO` built once outside that retry loop, so the first attempt drains it, and any retry after a transient failure silently uploads an empty file instead of actually failing.

Also not sure this fixes the original symptom. I traced `_to_v2_file_upload_item` in slack_sdk at 3.35.0 (pinned when this was filed) and current 3.43.0, and both turn `bytes` and `BytesIO` into identical `data` before upload, so I'm not sure what's different for Slack's rendering.

A fresh `BytesIO` per attempt would close the retry gap. Might be worth confirming the original bug still reproduces too before this lands.
